sdks/python/apache_beam/io/filesystem.py

```diff
@@ -0,0 +1,406 @@
+
+  def __hash__(self):
+    return hash((self.path, self.size_in_bytes))
+
+  def __ne__(self, other):
+    return not self.__eq__(other)
+
+  def __repr__(self):
+    return 'FileMetadata(%s, %s)' % (self.path, self.size_in_bytes)
+
+
+class FileSystem(object):
+  """A class that defines the functions that can be performed on a filesystem.
+
+  All methods are abstract and they are for file system providers to
+  implement. Clients should use the FileSystemUtil class to interact with
+  the correct file system based on the provided file pattern scheme.
+  """
+  __metaclass__ = abc.ABCMeta
+
+  @staticmethod
+  def _get_compression_type(path, compression_type):
+    if compression_type == CompressionTypes.AUTO:
+      compression_type = CompressionTypes.detect_compression_type(path)
+    elif not CompressionTypes.is_valid_compression_type(compression_type):
+      raise TypeError('compression_type must be CompressionType object but '
+                      'was %s' % type(compression_type))
+    return compression_type
+
+  @abc.abstractmethod
+  def mkdirs(self, path):
+    """Recursively create directories for the provided path.
+
+    Args:
+      path: string path of the directory structure that should be created
+
+    Raises:
+      IOError if leaf directory already exists.
+    """
+    raise NotImplementedError
+
+  @abc.abstractmethod
+  def match(self, patterns, limits=None):
+    """Find all matching paths to the patterns provided.
+
+    Args:
+      patterns: list of string for the file path pattern to match against
+      limits: list of maximum number of responses that need to be fetched
+
+    Returns: list of list of ``FileMetadata`` objects that match the patterns.
+    """
+    raise NotImplementedError
+
+  @abc.abstractmethod
+  def create(self, path, mime_type, compression_type):
+    """Returns a write channel for the given file path.
+
+    Args:
+      path: string path of the file object to be written to the system
+      mime_type: MIME type to specify the type on content in the file object
+      compression_type: Type of compression to be used for this object
+    """
+    raise NotImplementedError
+
+  @abc.abstractmethod
+  def open(self, path, mime_type, compression_type):
+    """Returns a read channel for the given file path.
+
+    Args:
+      path: string path of the file object to be written to the system
+      mime_type: MIME type to specify the type on content in the file object
+      compression_type: Type of compression to be used for this object
+    """
+    raise NotImplementedError
+
+  @abc.abstractmethod
+  def copy(self, sources, destinations):
+    """Recursively copy the file tree from the source to the destination
+
+    Args:
+      sources: list of source file/directory object that needs to be copied
+      destinations: list of destination of the new object
+    """
+    raise NotImplementedError
+
+  @abc.abstractmethod
+  def rename(self, sources, destinations):
```

Comment:
Please use source_resource_id(s) and destination_resource_id(s) in here and other places of FileSystems class (similar to Java SDK).